File: composeApp/src/androidMain/kotlin/com/aus/notelikeus/data/attachments/AttachmentBytesCodec.kt
package com.aus.notelikeus.data.attachments

import javax.crypto.Cipher
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

/**
 * On-disk attachment blob format: `NLA1` || IV(12) || ciphertext+tag.
 *
 * The provider generates the IV (AndroidKeyStore keys require randomized encryption). AAD binds
 * the ciphertext to a path identity so a swapped file fails authentication rather than silently
 * decoding under the wrong name.
 */
internal object AttachmentBytesCodec {
    private val MAGIC = byteArrayOf('N'.code.toByte(), 'L'.code.toByte(), 'A'.code.toByte(), '1'.code.toByte())
    private const val TRANSFORMATION = "AES/GCM/NoPadding"
    private const val IV_SIZE = 12
    private const val GCM_TAG_BITS = 128
    private const val MIN_SEALED = 4 + IV_SIZE + 16 // magic + iv + empty-tag minimum

    fun looksSealed(payload: ByteArray): Boolean =
        payload.size >= MIN_SEALED && payload.copyOfRange(0, MAGIC.size).contentEquals(MAGIC)

    fun seal(key: SecretKey, plaintext: ByteArray, aad: ByteArray): ByteArray {
        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(Cipher.ENCRYPT_MODE, key)
        if (aad.isNotEmpty()) cipher.updateAAD(aad)
        val iv = cipher.iv
        require(iv.size == IV_SIZE) { "Unexpected GCM IV size: ${iv.size}" }
        val ciphertext = cipher.doFinal(plaintext)
        val out = ByteArray(MAGIC.size + iv.size + ciphertext.size)
        System.arraycopy(MAGIC, 0, out, 0, MAGIC.size)
        System.arraycopy(iv, 0, out, MAGIC.size, iv.size)
        System.arraycopy(ciphertext, 0, out, MAGIC.size + iv.size, ciphertext.size)
        return out
    }

    fun open(key: SecretKey, payload: ByteArray, aad: ByteArray): ByteArray {
        require(looksSealed(payload)) { "Not a sealed attachment blob" }
        val iv = payload.copyOfRange(MAGIC.size, MAGIC.size + IV_SIZE)
        val ciphertext = payload.copyOfRange(MAGIC.size + IV_SIZE, payload.size)
        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(GCM_TAG_BITS, iv))
        if (aad.isNotEmpty()) cipher.updateAAD(aad)
        return cipher.doFinal(ciphertext)
    }
}

File: web/src/lib/crypto/attachmentBytesCodec.ts
/**
 * On-disk (IndexedDB) attachment blob format matching Android/Desktop:
 * `NLA1` || IV(12) || ciphertext+tag (AES-GCM-256).
 *
 * AAD binds the ciphertext to `ownerId/attachmentId` so a swapped blob fails authentication.
 */

const MAGIC = new Uint8Array([0x4e, 0x4c, 0x41, 0x31]); // NLA1
const IV_SIZE = 12;
const GCM_TAG_BYTES = 16;
const MIN_SEALED = MAGIC.length + IV_SIZE + GCM_TAG_BYTES;

export function attachmentAad(ownerId: string, attachmentId: string): Uint8Array {
  return new TextEncoder().encode(`${ownerId}/${attachmentId}`);
}

export function looksSealed(payload: Uint8Array): boolean {
  if (payload.byteLength < MIN_SEALED) return false;
  for (let i = 0; i < MAGIC.length; i++) {
    if (payload[i] !== MAGIC[i]) return false;
  }
  return true;
}

export async function sealAttachmentBytes(
  key: CryptoKey,
  plaintext: Uint8Array,
  aad: Uint8Array,
): Promise<Uint8Array> {
  const iv = crypto.getRandomValues(new Uint8Array(IV_SIZE));
  const ciphertext = new Uint8Array(
    await crypto.subtle.encrypt(
      {
        name: 'AES-GCM',
        iv,
        additionalData: aad,
      },
      key,
      plaintext,
    ),
  );
  const out = new Uint8Array(MAGIC.length + iv.length + ciphertext.length);
  out.set(MAGIC, 0);
  out.set(iv, MAGIC.length);
  out.set(ciphertext, MAGIC.length + iv.length);
  return out;
}

export async function openAttachmentBytes(
  key: CryptoKey,
  payload: Uint8Array,
  aad: Uint8Array,
): Promise<Uint8Array> {
  if (!looksSealed(payload)) {
    throw new Error('Not a sealed attachment blob');
  }
  const iv = payload.subarray(MAGIC.length, MAGIC.length + IV_SIZE);
  const ciphertext = payload.subarray(MAGIC.length + IV_SIZE);
  return new Uint8Array(
    await crypto.subtle.decrypt(
      {
        name: 'AES-GCM',
        iv,
        additionalData: aad,
      },
      key,
      ciphertext,
    ),
  );
}
